source/include/dqm4hep/Directory.h

```diff
@@ -140,21 +143,46 @@ namespace dqm4hep {
       /** Whether the directory is empty
        */
       bool isEmpty() const;
+      
+    private:
+      /** Default constructor
+       */
+      Directory();
 
+      /** Constructor with name and parent dir
+       */
+      Directory(const std::string &name, DirectoryPtr parentDir = nullptr);
+      
     private:
       std::string                     m_name;
-      Directory                      *m_pParent;
-      std::vector<Directory<T>*>      m_subdirs;
+      DirectoryPtr                    m_parent;
+      DirectoryList                   m_subdirs;
       ObjectList                      m_contents;
     };
 
     //-------------------------------------------------------------------------------------------------
     //-------------------------------------------------------------------------------------------------
 
+    template <typename T>
+    inline typename Directory<T>::DirectoryPtr Directory<T>::make_shared()
+    {
+      return std::shared_ptr<Directory<T>>(new Directory<T>());
```

Comment:
Shouldn't it be something like `return std::make_shared<Directory<T>>();` ?

Reply:
Normally yes, but the constructor is private to avoid raw pointer allocation. Thus the function make_shared can not call the constructor by itself.